src/Workspaces/Core/MSBuild/Rpc/RpcClient.cs

```diff
@@ -0,0 +1,150 @@
+                }
+            }
+
+            // We've disconnected, so cancel any remaining outstanding tasks
+            foreach (var (request, _) in _outstandingRequests.Values)
+                request.SetException(new System.Exception("The server disconnected unexpectedly."));
+
+            Disconnected?.Invoke(this, EventArgs.Empty);
+        });
+    }
+
+    public void Shutdown()
+    {
+        _shutdownTokenSource.Cancel();
+    }
+
+    public Task InvokeAsync(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken)
+    {
+        // We'll ignore the return value of InvokeCoreAsync in this case, since it'll just be filled in with null
+        return InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: null, cancellationToken);
+    }
+
+    public async Task<T?> InvokeNullableAsync<T>(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken) where T : class
+    {
+        var result = await InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: typeof(T), cancellationToken).ConfigureAwait(false);
+        return (T?)result;
+    }
+
+    public async Task<T> InvokeAsync<T>(int targetObject, string methodName, List<object?> parameters, CancellationToken cancellationToken) where T : notnull
+    {
+        var result = await InvokeCoreAsync(targetObject, methodName, parameters, expectedReturnType: typeof(T), cancellationToken).ConfigureAwait(false);
+        Contract.ThrowIfNull(result, "We expected a non-null result but got null back.");
+        return (T)result;
+    }
+
+    private async Task<object?> InvokeCoreAsync(int targetObject, string methodName, List<object?> parameters, Type? expectedReturnType, CancellationToken cancellationToken)
+    {
+        int requestId;
+        var requestCompletionSource = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        // Loop until we successfully have a new index for this; practically we don't expect this to ever collide, since that'd mean we'd have
+        // billions of requests outstanding, but...
+        do
+        {
+            requestId = Interlocked.Increment(ref _nextRequestId);
+        }
+        while (!_outstandingRequests.TryAdd(requestId, (requestCompletionSource, expectedReturnType)));
+
+        var request = new Request
+        {
+            Id = requestId,
+            TargetObject = targetObject,
+            Method = methodName,
+            Parameters = parameters.SelectAsArray(static p => p is not null ? JToken.FromObject(p) : JValue.CreateNull())
+        };
+
+        var requestJson = JsonConvert.SerializeObject(request, JsonSettings.SingleLineSerializerSettings);
+
+        using (await _sendingStreamSemaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+        {
+            _sendingStream.WriteLine(requestJson);
+            _sendingStream.Flush();
```

Comment:
what ensures the stream doesn't block?

Reply:
If your question is "why is this not WriteLineAsync?" I'll switch it. The stream of course could theoretically get blocked up if the other side isn't reading fast enough, but that's always a problem no matter what here. But since nothing else uses the stream other than this write location, we'll just be queuing the writes (the rest of the method is async) and they'll get there eventually.